fatrace.c

```diff
@@ -454,16 +469,71 @@ print_event (const struct fanotify_event_metadata *data,
                 problems[problem_idx++] = "path contains invalid UTF-8, path_raw contains the bytes.";
             }
         }
-        if (problem_idx) {
-          for (int i = 0; i < problem_idx; i++) {
-            printf("%s%s", i ? "\",\"" : ",\"problems\":[\"", problems[i]);
-          }
-          printf("\"]");
-        }
-        printf("}\n");
     } else {
-        printf ("%s(%i)%s: %-3s %s\n", procname[0] == '\0' ? "unknown" : procname, data->pid, printbuf, mask2str (data->mask), pathname);
+        printf ("%s(%i)%s: %-3s %s", procname[0] == '\0' ? "unknown" : procname, data->pid, printbuf, mask2str (data->mask), pathname);
+    }
+    if (option_ancestors && got_ppid) {
```

Comment:
This is a huge and complex loop and too hard to verify for a human. Please try to factorize this a bit to make it understandable, and use some helpers that open a file and read the contents safely.

Reply:
Good enough now IMHO.